# A1 Hindsight - Harrison Hauff - 48832061

## Objectives of this Assignment

#### In this Assignment I will be performing Data Analysis on a Structured Dataset. This will be done through addressing 6 tasks within the dataset.

### Tasks

1. Data Preperation
2. Exploratory Data Analysis
3. Basic Descriptive Statistics Analysis
4. Determining Correlations
5. Clustering
6. Principal Component Analysis

## Task 1: Data Preperation

To begin the project I will first read the dataset, utilising the pandas python library, and determining whether there is any missing data within the dataset.

In [58]:
import pandas as pd

bean_information = pd.read_csv("Drybean_dataset_2025.csv")

bean_information.head()

Area  Perimeter  MajorAxisLength  MinorAxisLength  AspectRation  \
0  28395    610.291       208.178117       173.888747      1.197191   
1  28734    638.018       200.524796       182.734419      1.097356   
2  29380    624.110       212.826130       175.931143      1.209713   
3  30008    645.884       210.557999       182.516516      1.153638   
4  30140    620.134       201.847882       190.279279      1.060798   

   Eccentricity  ConvexArea  EquivDiameter    Extent  Solidity    roundness  \
0      0.549812     28715.0     190.141097  0.763923  0.988856  0.958027126   
1      0.411785     29172.0     191.272751  0.783968  0.984986  0.887033637   
2      0.562727     29690.0     193.410904  0.778113  0.989559  0.947849473   
3      0.498616     30724.0     195.467062  0.782681  0.976696  0.903936374   
4      0.333680     30417.0     195.896503  0.773098  0.990893  0.984877069   

   Compactness  ShapeFactor1  ShapeFactor2  ShapeFactor3  ShapeFactor4  Class  
0     0.913358      0.007332      0.003147      0.834222      0.998724  SEKER  
1     0.953861      0.006979      0.003564      0.909851      0.998430  SEKER  
2     0.908774      0.007244      0.003048      0.825871      0.999066  SEKER  
3     0.928329      0.007017      0.003215      0.861794      0.994199  SEKER  
4     0.970516      0.006697      0.003665      0.941900      0.999166  SEKER

In [59]:
bean_information.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3871 entries, 0 to 3870
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Area             3860 non-null   object 
 1   Perimeter        3870 non-null   float64
 2   MajorAxisLength  3869 non-null   float64
 3   MinorAxisLength  3870 non-null   float64
 4   AspectRation     3869 non-null   float64
 5   Eccentricity     3871 non-null   float64
 6   ConvexArea       3870 non-null   float64
 7   EquivDiameter    3867 non-null   float64
 8   Extent           3870 non-null   float64
 9   Solidity         3870 non-null   float64
 10  roundness        3871 non-null   object 
 11  Compactness      3870 non-null   float64
 12  ShapeFactor1     3870 non-null   float64
 13  ShapeFactor2     3870 non-null   float64
 14  ShapeFactor3     3869 non-null   float64
 15  ShapeFactor4     3870 non-null   float64
 16  Class            3870 non-null   object 
dtypes: float64(14)

Given we have 3871 entries, we can see that some attributes have missing values, to get an exact count we'll utilise pandas to sum up the total amount of times where an entry is null. 

In [60]:
bean_information.isna().sum()

Area               11
Perimeter           1
MajorAxisLength     2
MinorAxisLength     1
AspectRation        2
Eccentricity        0
ConvexArea          1
EquivDiameter       4
Extent              1
Solidity            1
roundness           0
Compactness         1
ShapeFactor1        1
ShapeFactor2        1
ShapeFactor3        2
ShapeFactor4        1
Class               1
dtype: int64

As evident above, we do have characteristics missing only a few entries, while Eccentricity and Roundness both have no missing entries. Although these columns are only missing a few entries, I thought it would be appropriate to impute the missing data using the known data, doing this preserves the rows while maintaining a fair estimate which can still be insightful for our data analysis. 

First we'll compute the mean of the non-missing values for Area as it has the most missing values of any attribute in the dataset. Doing this requires us to convert the Area values from a string to an integer. 

In [61]:
# Convert to numeric safely
bean_information["Area"] = pd.to_numeric(bean_information["Area"], errors="coerce")

# Compute mean and round it
non_missing_area_mean = round(bean_information["Area"].mean())

# Fill missing values with the rounded mean
bean_information["Area"] = bean_information["Area"].fillna(non_missing_area_mean)


We'll check if imputing these values was successful in removing any empty columns. 

In [62]:
# verify that missing values are imputed in Area
bean_information.isna().sum()

Area               0
Perimeter          1
MajorAxisLength    2
MinorAxisLength    1
AspectRation       2
Eccentricity       0
ConvexArea         1
EquivDiameter      4
Extent             1
Solidity           1
roundness          0
Compactness        1
ShapeFactor1       1
ShapeFactor2       1
ShapeFactor3       2
ShapeFactor4       1
Class              1
dtype: int64

If we apply this to the rest of the attributes in the dataset.

#### Perimeter

In [63]:
bean_information["Perimeter"] = pd.to_numeric(bean_information["Perimeter"], errors="coerce")

non_missing_perimeter_mean = bean_information["Perimeter"].mean()

bean_information["Perimeter"] = bean_information["Perimeter"].fillna(non_missing_perimeter_mean)

#### Major Axis Length

In [67]:
bean_information["MajorAxisLength"] = pd.to_numeric(bean_information["MajorAxisLength"], errors="coerce")

non_missing_MAL_mean = bean_information["MajorAxisLength"].mean()

bean_information["MajorAxisLength"] = bean_information["MajorAxisLength"].fillna(non_missing_MAL_mean)

#### Minor Axis Length

In [68]:
bean_information["MinorAxisLength"] = pd.to_numeric(bean_information["MinorAxisLength"], errors="coerce")

non_missing_MAL_mean = bean_information["MinorAxisLength"].mean()

bean_information["MinorAxisLength"] = bean_information["MinorAxisLength"].fillna(non_missing_MAL_mean)

#### Aspect Ratio n

In [69]:
bean_information["AspectRation"] = pd.to_numeric(bean_information["AspectRation"], errors="coerce")

non_missing_ASN_mean = bean_information["AspectRation"].mean()

bean_information["AspectRation"] = bean_information["AspectRation"].fillna(non_missing_ASN_mean)

#### Convex Area

In [70]:
bean_information["ConvexArea"] = pd.to_numeric(bean_information["ConvexArea"], errors="coerce")

non_missing_CA_mean = bean_information["ConvexArea"].mean()

bean_information["ConvexArea"] = bean_information["ConvexArea"].fillna(non_missing_CA_mean)

#### Equiv Diameter

In [71]:
bean_information["EquivDiameter"] = pd.to_numeric(bean_information["EquivDiameter"], errors="coerce")

non_missing_ED_mean = bean_information["EquivDiameter"].mean()

bean_information["EquivDiameter"] = bean_information["EquivDiameter"].fillna(non_missing_ED_mean)

#### Extent

In [72]:
bean_information["Extent"] = pd.to_numeric(bean_information["Extent"], errors="coerce")

non_missing_extent_mean = bean_information["Extent"].mean()

bean_information["Extent"] = bean_information["Extent"].fillna(non_missing_extent_mean)

#### Solidity

In [73]:
bean_information["Solidity"] = pd.to_numeric(bean_information["Solidity"], errors="coerce")

non_missing_solidity_mean = bean_information["Solidity"].mean()

bean_information["Solidity"] = bean_information["Solidity"].fillna(non_missing_solidity_mean)

#### Compactness

In [74]:
bean_information["Compactness"] = pd.to_numeric(bean_information["Compactness"], errors="coerce")

non_missing_compactness_mean = bean_information["Compactness"].mean()

bean_information["Compactness"] = bean_information["Compactness"].fillna(non_missing_compactness_mean)

#### Shape Factor's 1, 2, 3, 4

In [75]:
# Shape Factor 1
bean_information["ShapeFactor1"] = pd.to_numeric(bean_information["ShapeFactor1"], errors="coerce")

non_missing_SF1_mean = bean_information["ShapeFactor1"].mean()

bean_information["ShapeFactor1"] = bean_information["ShapeFactor1"].fillna(non_missing_SF1_mean)

# Shape Factor 2

bean_information["ShapeFactor2"] = pd.to_numeric(bean_information["ShapeFactor2"], errors="coerce")

non_missing_SF2_mean = bean_information["ShapeFactor2"].mean()

bean_information["ShapeFactor2"] = bean_information["ShapeFactor2"].fillna(non_missing_SF2_mean)

# Shape Factor 3

bean_information["ShapeFactor3"] = pd.to_numeric(bean_information["ShapeFactor3"], errors="coerce")

non_missing_SF3_mean = bean_information["ShapeFactor3"].mean()

bean_information["ShapeFactor3"] = bean_information["ShapeFactor3"].fillna(non_missing_SF3_mean)

# Shape Factor 4

bean_information["ShapeFactor4"] = pd.to_numeric(bean_information["ShapeFactor4"], errors="coerce")

non_missing_SF4_mean = bean_information["ShapeFactor4"].mean()

bean_information["ShapeFactor4"] = bean_information["ShapeFactor4"].fillna(non_missing_SF4_mean)



In [76]:
bean_information.isna().sum()

Area               0
Perimeter          0
MajorAxisLength    0
MinorAxisLength    0
AspectRation       0
Eccentricity       0
ConvexArea         0
EquivDiameter      0
Extent             0
Solidity           0
roundness          0
Compactness        0
ShapeFactor1       0
ShapeFactor2       0
ShapeFactor3       0
ShapeFactor4       0
Class              1
dtype: int64

Given that class is not a numerical data type and is fact a categorical type, I decided it was necessary to delete this row as a mean estimation cannot be made regarding its value, and additionally it is only 1 row and will not have massive ramifications on the information gained from the data. 

#### Class

In [78]:
bean_information = bean_information[bean_information['Class'].notna()]

#### Final Cleaned Dataset

In [79]:
bean_information.isna().sum()

Area               0
Perimeter          0
MajorAxisLength    0
MinorAxisLength    0
AspectRation       0
Eccentricity       0
ConvexArea         0
EquivDiameter      0
Extent             0
Solidity           0
roundness          0
Compactness        0
ShapeFactor1       0
ShapeFactor2       0
ShapeFactor3       0
ShapeFactor4       0
Class              0
dtype: int64

## Task 2: Exploratory Data Analysis

The purpose of exploratory data analysis (EDA) is to gather an understanding of the datas main characteristics through the use of descriptive statistics and data visualisation.

### Univariate EDA

Univariate analysis explores columns/characteristics in isolation in order to ascertain their distribution.

Since we have already have dealt with missing values, we can use the univariate EDA to determine duplicates or outliers within the dataset

### Area Univariate EDA

In [83]:
bean_information["Area"].describe()

count      3870.000000
mean      68051.156072
std       45042.046213
min       28395.000000
25%       38948.500000
50%       48384.500000
75%       73233.000000
max      254616.000000
Name: Area, dtype: float64

It can be observed from the results above that we may have a signifficant skew in our area values, this is due to the mean and median being signifficantly different, with the mean being 68051.16 and median being 48384.5. Additionally if we use the formula for z scores, we can determine how far the max and min values are from the mean based on the standard deviation.  

In [84]:
area_mean = 68051.156072
area_std = 45042.046213
area_max = 254616.00

max_z_score = ((area_max - area_mean)/area_std)
print(max_z_score)

4.142015286022992


If we compare this to the min area from the same dataset.

In [85]:
area_mean = 68051.156072
area_std = 45042.046213
area_min = 28395.000000

min_z_score = ((area_mean - area_min)/area_std)
print(min_z_score)

0.8804252782937394


With the max z score being 4.14 standard deviations away from the mean, and the min only being 0.88 standard deviations away, it is clear there is a clear data quality issue present. If we count the amount of entries above a certain threshold we can determine if this disparity is due to an issue or a characteristic of the dataset.

In [90]:
# say we set the threshold to 3 standard deviations above the mean to start with
area_mean = 68051.156072
area_std = 45042.046213
threshold = area_mean + 3*area_std

outliers = (bean_information['Area'] > threshold).sum()
print(outliers)

60


### Perimeter Univariate EDA